### ISSUE TITLE
Prometheus message resource limit check performance issues

### ISSUE THREAD

**Issue by bordeuax** [eclipse/hono#2084]
**Production Problem**
Daily in production i am facing with regular CPU/memory spikes for Prometheus, as a results of such behavior we have messaging/connection timeouts and huge performance degradation of message resource limit check. 
**Particular use case**
A tenant pushing events/telemetry/command each 15 minutes via HTTP (i think this should be available for other adapters) number of devices  that doing connections are around 30-40. Prometheus cache timeout is 600s, 3 instances of HTTP adapter and 4 CPU core per pod (4 verticles). Prometheus has been configured to log all the queries into a file , for more info see [query log](https://prometheus.io/docs/guides/query-log/). 
**Prometheus query log**
```
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.239Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.239Z","step":0},"stats":{"timings":{"evalTotalTime":3.700122281,"resultSortTime":0,"queryPreparationTime":0.002982689,"innerEvalTime":3.697125523,"execQueueTime":0.000013985,"execTotalTime":3.700148945}},"ts":"2020-07-15T09:30:03.939Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.431Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.431Z","step":0},"stats":{"timings":{"evalTotalTime":3.579733563,"resultSortTime":0,"queryPreparationTime":0.143595289,"innerEvalTime":3.43611965,"execQueueTime":0.000020175,"execTotalTime":3.57976751}},"ts":"2020-07-15T09:30:04.011Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.373Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.373Z","step":0},"stats":{"timings":{"evalTotalTime":3.649874872,"resultSortTime":0,"queryPreparationTime":0.007993198,"innerEvalTime":3.641865574,"execQueueTime":0.000014219,"execTotalTime":3.6498991480000003}},"ts":"2020-07-15T09:30:04.023Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.333Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.333Z","step":0},"stats":{"timings":{"evalTotalTime":3.709548315,"resultSortTime":0,"queryPreparationTime":0.00271979,"innerEvalTime":3.70681296,"execQueueTime":0.000010563,"execTotalTime":3.709569187}},"ts":"2020-07-15T09:30:04.043Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.491Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.491Z","step":0},"stats":{"timings":{"evalTotalTime":3.5879867130000003,"resultSortTime":0,"queryPreparationTime":0.008250753,"innerEvalTime":3.579718649,"execQueueTime":0.000013822,"execTotalTime":3.58801237}},"ts":"2020-07-15T09:30:04.079Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.388Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.388Z","step":0},"stats":{"timings":{"evalTotalTime":3.703537024,"resultSortTime":0,"queryPreparationTime":0.002860165,"innerEvalTime":3.700662124,"execQueueTime":0.000014081,"execTotalTime":3.70356104}},"ts":"2020-07-15T09:30:04.092Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.368Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.368Z","step":0},"stats":{"timings":{"evalTotalTime":3.726635575,"resultSortTime":0,"queryPreparationTime":0.003062827,"innerEvalTime":3.723557226,"execQueueTime":0.000013874,"execTotalTime":3.726663855}},"ts":"2020-07-15T09:30:04.095Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.239Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.239Z","step":0},"stats":{"timings":{"evalTotalTime":4.022926031,"resultSortTime":0,"queryPreparationTime":0.00298721,"innerEvalTime":4.019923616,"execQueueTime":0.000016764,"execTotalTime":4.022955417}},"ts":"2020-07-15T09:30:04.262Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.546Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.546Z","step":0},"stats":{"timings":{"evalTotalTime":3.744322794,"resultSortTime":0,"queryPreparationTime":0.015392899,"innerEvalTime":3.728910452,"execQueueTime":0.000014263,"execTotalTime":3.744349555}},"ts":"2020-07-15T09:30:04.295Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.508Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.508Z","step":0},"stats":{"timings":{"evalTotalTime":3.911230487,"resultSortTime":0,"queryPreparationTime":0.01599865,"innerEvalTime":3.895031167,"execQueueTime":0.000014099,"execTotalTime":3.9112562889999998}},"ts":"2020-07-15T09:30:04.420Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.411Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.411Z","step":0},"stats":{"timings":{"evalTotalTime":4.063279805,"resultSortTime":0,"queryPreparationTime":0.007814643,"innerEvalTime":4.055446178,"execQueueTime":0.000014697,"execTotalTime":4.06330686}},"ts":"2020-07-15T09:30:04.475Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.451Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.451Z","step":0},"stats":{"timings":{"evalTotalTime":4.046240517,"resultSortTime":0,"queryPreparationTime":0.015948872,"innerEvalTime":4.030271228,"execQueueTime":0.000017049,"execTotalTime":4.046269743}},"ts":"2020-07-15T09:30:04.498Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:00.617Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:00.617Z","step":0},"stats":{"timings":{"evalTotalTime":3.899746093,"resultSortTime":0,"queryPreparationTime":0.005566466,"innerEvalTime":3.894162581,"execQueueTime":0.000021842,"execTotalTime":3.899781999}},"ts":"2020-07-15T09:30:04.518Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:02.946Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:02.946Z","step":0},"stats":{"timings":{"evalTotalTime":2.877515293,"resultSortTime":0,"queryPreparationTime":0.003240674,"innerEvalTime":2.874258438,"execQueueTime":0.000019004,"execTotalTime":2.877545638}},"ts":"2020-07-15T09:30:05.824Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:03.941Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:03.941Z","step":0},"stats":{"timings":{"evalTotalTime":2.332415507,"resultSortTime":0,"queryPreparationTime":0.003132921,"innerEvalTime":2.329268073,"execQueueTime":0.000014178,"execTotalTime":2.332440435}},"ts":"2020-07-15T09:30:06.273Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:03.300Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:03.300Z","step":0},"stats":{"timings":{"evalTotalTime":3.146027013,"resultSortTime":0,"queryPreparationTime":0.004561806,"innerEvalTime":3.141447013,"execQueueTime":0.000018313,"execTotalTime":3.146063201}},"ts":"2020-07-15T09:30:06.447Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.171Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.171Z","step":0},"stats":{"timings":{"evalTotalTime":2.869228611,"resultSortTime":0,"queryPreparationTime":0.005204836,"innerEvalTime":2.864007968,"execQueueTime":0.000011908,"execTotalTime":2.869252165}},"ts":"2020-07-15T09:30:07.040Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.092Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.092Z","step":0},"stats":{"timings":{"evalTotalTime":3.091659182,"resultSortTime":0,"queryPreparationTime":0.003179361,"innerEvalTime":3.088464164,"execQueueTime":0.000012639,"execTotalTime":3.091691836}},"ts":"2020-07-15T09:30:07.184Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.110Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.110Z","step":0},"stats":{"timings":{"evalTotalTime":3.074048838,"resultSortTime":0,"queryPreparationTime":0.003038825,"innerEvalTime":3.070996591,"execQueueTime":0.00001206,"execTotalTime":3.074071476}},"ts":"2020-07-15T09:30:07.185Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.138Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.138Z","step":0},"stats":{"timings":{"evalTotalTime":3.054557692,"resultSortTime":0,"queryPreparationTime":0.005741075,"innerEvalTime":3.048796245,"execQueueTime":0.000035824,"execTotalTime":3.054602984}},"ts":"2020-07-15T09:30:07.193Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.342Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.342Z","step":0},"stats":{"timings":{"evalTotalTime":3.065535229,"resultSortTime":0,"queryPreparationTime":0.003538705,"innerEvalTime":3.061977503,"execQueueTime":0.000016778,"execTotalTime":3.065564342}},"ts":"2020-07-15T09:30:07.408Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.151Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.151Z","step":0},"stats":{"timings":{"evalTotalTime":3.277817416,"resultSortTime":0,"queryPreparationTime":0.002959559,"innerEvalTime":3.274841895,"execQueueTime":0.000012214,"execTotalTime":3.277841584}},"ts":"2020-07-15T09:30:07.429Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.312Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.312Z","step":0},"stats":{"timings":{"evalTotalTime":3.249605775,"resultSortTime":0,"queryPreparationTime":0.003491823,"innerEvalTime":3.246095773,"execQueueTime":0.000014001,"execTotalTime":3.249641789}},"ts":"2020-07-15T09:30:07.562Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.060Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.060Z","step":0},"stats":{"timings":{"evalTotalTime":3.570850685,"resultSortTime":0,"queryPreparationTime":0.003133972,"innerEvalTime":3.567698648,"execQueueTime":0.000014741,"execTotalTime":3.570879215}},"ts":"2020-07-15T09:30:07.631Z"}
{"httpRequest":{"clientIP":"100.100.0.22","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.447Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.447Z","step":0},"stats":{"timings":{"evalTotalTime":3.462727139,"resultSortTime":0,"queryPreparationTime":0.003321888,"innerEvalTime":3.459387881,"execQueueTime":0.000013791,"execTotalTime":3.46275269}},"ts":"2020-07-15T09:30:07.910Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.507Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.507Z","step":0},"stats":{"timings":{"evalTotalTime":3.412886102,"resultSortTime":0,"queryPreparationTime":0.003338769,"innerEvalTime":3.409531728,"execQueueTime":0.000015544,"execTotalTime":3.4129127329999998}},"ts":"2020-07-15T09:30:07.920Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.568Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.568Z","step":0},"stats":{"timings":{"evalTotalTime":3.371724936,"resultSortTime":0,"queryPreparationTime":0.003144912,"innerEvalTime":3.368566419,"execQueueTime":0.000014043,"execTotalTime":3.371749504}},"ts":"2020-07-15T09:30:07.940Z"}
{"httpRequest":{"clientIP":"100.108.0.17","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:04.558Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:04.558Z","step":0},"stats":{"timings":{"evalTotalTime":3.394222939,"resultSortTime":0,"queryPreparationTime":0.003353455,"innerEvalTime":3.390852959,"execQueueTime":0.000014908,"execTotalTime":3.394248232}},"ts":"2020-07-15T09:30:07.952Z"}
{"httpRequest":{"clientIP":"100.124.0.11","method":"GET","path":"/api/v1/query"},"params":{"end":"2020-07-15T09:30:05.891Z","query":"floor(sum(increase(hono_messages_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_commands_payload_bytes_sum*0) + sum(increase(hono_commands_payload_bytes_sum{status=~\"forwarded|unprocessable\", tenant=\"tenant_id\"} [15d]) or hono_messages_payload_bytes_sum*0))","start":"2020-07-15T09:30:05.891Z","step":0},"stats":{"timings":{"evalTotalTime":2.155890866,"resultSortTime":0,"queryPreparationTime":0.003157182,"innerEvalTime":2.1527185,"execQueueTime":0.000016671,"execTotalTime":2.155917501}},"ts":"2020-07-15T09:30:08.047Z"}
```
 
**Suggestions** 
* From attached logs i see that we have a concurrency issue when one protocol adapter query "message limits" for tenant. 
* To have better performance i suggest to use  [Prometheus recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) or to implement the second  _resource limit check_ based on recording rules.

**Comment by sophokles73** [comment]
@bordeuax FMPOV there are multiple aspects to this.
How long does the execution of a single query actually take and how often is it executed. What I do not see from the logs is the unit of measure for the execution time. Is that seconds or milliseconds?
I had already been thinking about using a recording rule but did not (yet) figure out how that should work, given the dynamic aspect of the time range parameter. Any idea?

**Comment by sophokles73** [comment]
Ok, so if I am reading this correctly, then the *execTotalTime* and the *ts* properties indicate that the queries take about 3 seconds to execute (which is terrible). However, I do not (yet) see what kind of recording rule could do the trick. Maybe there is also another (more efficient) way of formulating the query itself ...

**Comment by bordeuax** [comment]
@sophokles73 , i see this implementation in the following way: 
* Create new resource limit check  based on recording rules
* Extends or give a possibility to have custom values in the field _mode_  from [Resource Limits Period Configuration](https://www.eclipse.org/hono/docs/api/tenant/#resource-limits-period-configuration-format). And this values will be interpreted by  rules resource checker as a name for prometheus recorded rule. 

Regarding dynamic aspect of time range parameter, here I see 2 possibility 
* when the _mod_ is _monthly_ try to using [day_of_month](https://prometheus.io/docs/prometheus/latest/querying/functions/#day_of_month) function 
* When the users have _days_ mod enable, we can create rules per time range; for example a rule like `tenant_messages _last_10d` will collect bytes for last 10 days and the time range (10 days) will be hard-coded in the rule. I don't see the need to have very fine grain rules per tenant, usually this rules are per service plan or package and create this rules will be on the responsibility of Hono operators.

**Comment by kaniyan** [comment]
> Maybe there is also another (more efficient) way of formulating the query itself ...

_Query formulation:_
The message limit query aggregates data from two different time series (`hono_messages_payload_bytes_sum` and `hono_commands_payload_bytes_sum`). With some existing preconditions (such as one time series may exist and other not), I don't see at the moment any other better way of writing this query. Also I checked this query in the same environment (as stated in this issue) and it takes around 500 milliseconds on average. As I understood, in general it doesn't take 3 seconds. I don't think rewriting this query would resolve this issue. Hence I think, its worthwhile to monitor the performance parameters of the Prometheus server during such spikes to understand what cause these spikes.

It is also possible to split the above query into two queries (as given below) and execute them in parallel. And each query took on average around 350 ms in the same environment. We could split those query into two and execute parallelly. IMHO it doesn't bring any tremendous improvement or resolve this issue.
```
floor(sum(increase(hono_messages_payload_bytes_sum{status=~"forwarded|unprocessable", tenant="DEFAULT_TENANT"} [15d])))
floor(sum(increase(hono_commands_payload_bytes_sum{status=~"forwarded|unprocessable", tenant="DEFAULT_TENANT"} [15d])))
```

_Concurrent execution:_
Let us assume that the cached entry corresponding to a tenant has expired and a device sends a telemetry message. Since there is no cached entry for the tenant, the protocol adapter sends a _HTTP_ request to the Prometheus server to execute the query. Meanwhile, before completion of the query execution, any messages that arrive would trigger _HTTP_ requests to the Prometheus server to execute the query. This could be avoided by ensuring that if a query execution for a tenant is in progress then the other messages must wait and not trigger further query executions.

**Comment by sophokles73** [comment]
@kaniyan the problem with concurrent execution is definitely worth solving :-) We need to make sure that we return the previous (expired) value in concurrent invocations of the method while a query to Prometheus is still executing. We should be able to use `Caffeine.refreshAfterWrite()` for the data volume cache instead of `Caffeine.expireAfterWrite()`. WDYT?

**Comment by sophokles73** [comment]
@bordeuax 

> Regarding dynamic aspect of time range parameter, here I see 2 possibility
>*    when the mod is monthly try to using day_of_month function

I do not quite see how the *day_of_month* function could be helpful here. Can you explain?

>*    When the users have days mod enable, we can create rules per time range; for example a rule like tenant_messages _last_10d will collect bytes for last 10 days and the time range (10 days) will be hard-coded in the rule. I don't see the need to have very fine grain rules per tenant, usually this rules are per service plan or package and create this rules will be on the responsibility of Hono operators.

Based on the assumption that a tenant's devices upload MB of data per hour, I wonder if rules using granularity of multiple days would be any help at all. FMPOV a useful granularity would be more in the minutes region ...

**Comment by sophokles73** [comment]
@kaniyan The Caffeine cache already takes care of atomically loading a value for a key that doesn't exist in the cache, i.e. the Prometheus query will be executed only once per missing key.

**Comment by sophokles73** [comment]
A possible option to improve the situation might be to apply the same approach as used for recording the number of connections, i.e. periodically reporting the accumulated number of bytes processed by the adapter.

**Comment by sophokles73** [comment]
It seems that the existing implementation is prone to generating a lot of load on a Prometheus server under certain conditions. In particular, the load would spike if a protocol adapter to which many devices of many different tenants were connected would crash (or be shut down) and the devices would immediately re-connect to another instance of the protocol adapter.

In this case, for the first device per tenant up to three queries would be issued to Prometheus' query API, depending on whether the tenant's current limited resource values were already available in the cache or not. For multiple devices from different tenants connecting in parallel, this could lead to a high load of queries on the Prometheus server.

The reason for this seems to be the ¸java.util.concurrent.Executor` returned by `java.util.concurrent.ForkJoinPool.commonPool()` which is used by default in the Caffeine cache for running tasks asynchronously, e.g. for computing values. The executor uses multiple threads for executing tasks which results in multiple queries being issued (almost) in parallel which in turn may result in the high load on the Prometheus server for executing these queries. The queries themselves seem to take 200-500ms each to execute on the server.

Another issue with the current implementation is that it *chains* the Futures returned by the check methods to the task retrieving a missing value for a tenant. IN the scenario described above, this results in all the devices' connection attempts waiting for the retrieval of the limited resource value into the cache. If the query takes a long time to run (due to the high load on the Prometheus server), then all the devices' connection attempts will be in vain. Instead, it seems to be more appropriate to simply pass all checks unless the cache contains the corresponding limited resource value. This also seams reasonable for the scenario described above because the devices that have lost their connections with the shut down adapter instance have been eligible to maintain their connection with the shut down adapter in the first place.
